### PR TITLE
Push instance-to-node mapping from clustermgtd to log file

### PR DESCRIPTION
**Pull request by Jennifer603**
### Description of changes
* Added a new logger nodemap
* Added a event publisher that uses the new logger
* Add a new publisher function to publish the instance_to_node information
* Added command in clustermgtd.py to call this new method

### Tests
* Created a unit test to test the publish_nodemapping_event
* The unit test contains situations: one node in nodemap does not have an instance, empty nodemap, and node map does not log warning.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/slurm_plugin/cluster_event_publisher.py [hunk 1]
```diff
@@ -509,6 +513,47 @@ def publish_node_launch_events(self, failed_nodes: Dict[str, List[str]]):
...
+    #     "scheduler": "Slurm",
+    #     "message": <str>,
+    #     "detail": {
+    #         # instance to node mapping schema
+    #         "node_list":[
+    #             {
+    #                 "instance_id":"i-123",
+    #                 "node_name": "string",
+    #                 "instance_type": "string",
+    #                 "disable_multi_threading": "bool"
```
Comment (delongmeng) [inline src/slurm_plugin/cluster_event_publisher.py]: should be `threads_per_core`?

File: src/slurm_plugin/cluster_event_publisher.py [hunk 2]
```diff
@@ -597,6 +642,21 @@ def _failed_node_supplier(unhealthy_static_nodes: List[SlurmNode], failed_nodes:
                         }
                     }
 
+    def _nodemapping_supplier(self, node_map: List[SlurmNode]) -> Iterator:
+        for node in node_map:
```
Comment (delongmeng) [inline src/slurm_plugin/cluster_event_publisher.py]: `node_map` -> should just be `nodes`? or something like `nodes_to_map`?

**Review by delongmeng — COMMENTED** [review]
LGTM
please merge into a feature branch instead of develop

File: src/slurm_plugin/cluster_event_publisher.py (hunk 1 above)
Comment (Jennifer603) [reply src/slurm_plugin/cluster_event_publisher.py]: Done!

File: src/slurm_plugin/cluster_event_publisher.py (hunk 2 above)
Comment (Jennifer603) [reply src/slurm_plugin/cluster_event_publisher.py]: Done!

File: src/slurm_plugin/cluster_event_publisher.py [hunk 3]
```diff
@@ -597,6 +642,21 @@ def _failed_node_supplier(unhealthy_static_nodes: List[SlurmNode], failed_nodes:
                         }
                     }
 
+    def _nodemapping_supplier(self, nodes: List[SlurmNode]) -> Iterator:
```
Comment (chenwany) [inline src/slurm_plugin/cluster_event_publisher.py]: Is this a static method? Could you add `@staticmethod`?

Comment (chenwany) [inline src/slurm_plugin/cluster_event_publisher.py]: could you separate nodemapping to node_mapping

File: src/slurm_plugin/cluster_event_publisher.py [hunk 4]
```diff
@@ -597,6 +642,21 @@ def _failed_node_supplier(unhealthy_static_nodes: List[SlurmNode], failed_nodes:
...
+        for node in nodes:
+            yield {
+                "detail": {
+                    "nodes": [
```
Comment (chenwany) [inline src/slurm_plugin/cluster_event_publisher.py]: With this we are creating one detail for each node and only have a single node in the `nodes` list , this is not what we expect to have.
Could you fix this by including all the nodes into the `nodes` list?

Comment (chenwany) [reply src/slurm_plugin/cluster_event_publisher.py]: After changing this, could you add an example test output of the json event in the PR description?

File: src/slurm_plugin/cluster_event_publisher.py (hunk 3 above)
Comment (Jennifer603) [reply src/slurm_plugin/cluster_event_publisher.py]: Done

**Review by davprat — DISMISSED** [review]
Looks good!

**Review by davprat — COMMENTED** [review]
Actually, you should have some formatting errors as reported by tox, so run tox -e autoformat and update your PR, then I can approve it.

File: src/slurm_plugin/cluster_event_publisher.py (hunk 4 above)
Comment (Jennifer603) [reply src/slurm_plugin/cluster_event_publisher.py]: done

**Pull request merged**